### PR TITLE
[NVIDIA] Add Cutlass MLA backend

### PR DESCRIPTION
**Pull request by kaixih**
This PR introduces the `CUTLASS_MLA_VLLM_V1` backend, enabling support for `ops.cutlass_mla_decode()` on NVIDIA Blackwell GPUs.

It also includes performance results using DeepSeek-V3 on 8×B200 GPUs under DP+EP parallelism settings, which delivers ~17% improved throughput.

```
# With default triton backend:
============ Serving Benchmark Result ============
Successful requests:                     2989
Benchmark duration (s):                  1046.01
Total input tokens:                      2989000
Total generated tokens:                  2989000
Request throughput (req/s):              2.86
Output token throughput (tok/s):         2857.52
Total Token throughput (tok/s):          5715.04
---------------Time to First Token----------------
Mean TTFT (ms):                          200716.51
Median TTFT (ms):                        199463.35
P99 TTFT (ms):                           395239.25
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          826.04
Median TPOT (ms):                        826.20
P99 TPOT (ms):                           1001.39
---------------Inter-token Latency----------------
Mean ITL (ms):                           826.04
Median ITL (ms):                         648.89
P99 ITL (ms):                            8337.69
==================================================

With cutlass_mla backend:
============ Serving Benchmark Result ============
Successful requests:                     2989
Benchmark duration (s):                  881.52
Total input tokens:                      2989000
Total generated tokens:                  2989000
Request throughput (req/s):              3.39
Output token throughput (tok/s):         3390.73
Total Token throughput (tok/s):          6781.46
---------------Time to First Token----------------
Mean TTFT (ms):                          190244.11
Median TTFT (ms):                        189563.96
P99 TTFT (ms):                           372713.07
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          685.60
Median TPOT (ms):                        686.96
P99 TPOT (ms):                           858.01
---------------Inter-token Latency----------------
Mean ITL (ms):                           685.60
Median ITL (ms):                         518.56
P99 ITL (ms):                            7738.23
==================================================
```

To repro the results:
```bash
# Server side with triton backend (Plz use VLLM_ATTENTION_BACKEND=CUTLASS_MLA_VLLM_V1 for cutlass backend):
VLLM_LOGGING_LEVEL=DEBUG \
VLLM_WORKER_MULTIPROC_METHOD=spawn \
  vllm serve deepseek-ai/DeepSeek-V3 \
    --trust-remote-code \
    --max-model-len=2048 \
    --block-size=128 \
    --max-num-seqs=512 \
    --gpu_memory_utilization=0.97 \
    --data-parallel-size $NUM_GPUS --enable-expert-parallel \
    --disable-log-requests

# client side:
python $VLLM_PATH/benchmarks/benchmark_serving.py \
  --model deepseek-ai/DeepSeek-V3 \
  --dataset-name random \
  --ignore-eos \
  --num-prompts 3000 \
  --max-concurrency 3000 \
  --random-input-len 1000 \
  --random-output-len 1000
```

cc. @kushanam

### CODE REVIEW

File: vllm/v1/attention/backends/mla/cutlass_mla.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+        if attn_type != AttentionType.DECODER:
+            raise NotImplementedError("Encoder self-attention and "
+                                      "encoder/decoder cross-attention "
+                                      "are not implemented for "
+                                      "TritonMLAImpl")
+
+        if is_quantized_kv_cache(self.kv_cache_dtype):
+            raise NotImplementedError(
+                "TritonMLA V1 with FP8 KV cache not yet supported")
```
[the remark this answers is not in the document]
Comment (kaixih) [reply vllm/v1/attention/backends/mla/cutlass_mla.py]: Done. PTAL.

File: vllm/v1/attention/backends/mla/common.py [hunk 2]
```diff
@@ -349,7 +349,7 @@ def __init__(self,
         self.num_heads = model_config.get_num_attention_heads(
             runner.parallel_config)
         self.mla_dims = get_mla_dims(model_config)
-        self.aot_schedule = is_vllm_fa and (get_flash_attn_version() == 3)
+        self.aot_schedule = current_platform.is_cuda()
 
         # Dont try to access the runner on AMD
         if self.aot_schedule:
```
Comment (LucasWilkinson) [inline vllm/v1/attention/backends/mla/common.py:355]: we should leave
```
self.aot_schedule = is_vllm_fa and (get_flash_attn_version() == 3)
```
and updated:
```
if current_platform.is_cuda():
     self.page_size = self.runner.block_size
```
to
```
if self.aot_schedule:
     self.page_size = self.runner.block_size
```

File: vllm/platforms/cuda.py [hunk 3]
```diff
@@ -179,6 +179,14 @@ def get_attn_backend_cls(cls, selected_backend, head_size, dtype,
         if use_mla:
             # TODO(lucas): refactor to  be more concise
             #  we should probably consider factoring out V1 here
+            if selected_backend == _Backend.CUTLASS_MLA_VLLM_V1:
```
Comment (LucasWilkinson) [inline vllm/platforms/cuda.py:182]: the perf looks really good! I think we should turn this on by default for blackwell

File: vllm/v1/attention/backends/mla/cutlass_mla.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    ) -> torch.Tensor:
+        assert kv_c_and_k_pe_cache.numel() > 0
+        assert attn_metadata.decode is not None
+
+        if self.kv_cache_dtype.startswith("fp8"):
+            raise NotImplementedError("FP8 Cutlass MLA not yet supported")
+
+        B = q_nope.shape[0]
+
+        o = torch.zeros(B,
```
Comment (LucasWilkinson) [inline vllm/v1/attention/backends/mla/cutlass_mla.py:84]: does this need to be `torch.zeros`? or does `torch.empty` work (faster)

**Review by LucasWilkinson — COMMENTED** [review]
The perf is looking really good! thanks for the contribution!

Do you mind doing accuracy checks 
```
VLLM_ATTENTION_BACKEND=CUTLASS_MLA_VLLM_V1 VLLM_USE_V1=1 lm-eval --model vllm --model_args pretrained=deepseek-ai/DeepSeek-V2-Lite-Chat,tensor_parallel_size=2,dtype=auto,gpu_memory_utilization=0.9,trust_remote_code=True,max_model_len=16384 --task gsm8k --num_fewshot 5  --batch_size auto
```

**Comment by kaixih** [comment]
> The perf is looking really good! thanks for the contribution!
> 
> Do you mind doing accuracy checks
> 
> ```
> VLLM_ATTENTION_BACKEND=CUTLASS_MLA_VLLM_V1 VLLM_USE_V1=1 lm-eval --model vllm --model_args pretrained=deepseek-ai/DeepSeek-V2-Lite-Chat,tensor_parallel_size=2,dtype=auto,gpu_memory_utilization=0.9,trust_remote_code=True,max_model_len=16384 --task gsm8k --num_fewshot 5  --batch_size auto
> ```

Do you know any commend to test a model with num_heads = 128? And probably no TP.

**Comment by LucasWilkinson** [comment]
> > ```
> > VLLM_ATTENTION_BACKEND=CUTLASS_MLA_VLLM_V1 VLLM_USE_V1=1 lm-eval --model vllm --model_args pretrained=deepseek-ai/DeepSeek-V2-Lite-Chat,tensor_parallel_size=2,dtype=auto,gpu_memory_utilization=0.9,trust_remote_code=True,max_model_len=16384 --task gsm8k --num_fewshot 5  --batch_size auto
> > ```

Not that im aware of :/ this is the smallest MLA model I am aware of

**Comment by mgoin** [comment]
This is the smallest model with MLA ( @tlrmchlsmth found it the other day) https://huggingface.co/deepseek-ai/deepseek-vl2-tiny

**Comment by tlrmchlsmth** [comment]
> This is the smallest model with MLA ( @tlrmchlsmth found it the other day) https://huggingface.co/deepseek-ai/deepseek-vl2-tiny

just want to clarify that the way I found it was `grep -r deepseek tests | grep tiny`

**Comment by LucasWilkinson** [comment]
Ah I don't think it's an MLA model :/
```
    "kv_lora_rank": null,
```

**Review by LucasWilkinson — APPROVED** [review]
LGTM, other than I do think we should turn it on by default for Blackwell: https://github.com/vllm-project/vllm/pull/17625/files#r2075575324

Any reason not to?

Edit: oh and ideally id still like to see accuracy numbers, i.e.:
```
VLLM_ATTENTION_BACKEND=CUTLASS_MLA_VLLM_V1 VLLM_USE_V1=1 lm-eval --model vllm --model_args pretrained=deepseek-ai/DeepSeek-V2-Lite-Chat,tensor_parallel_size=2,dtype=auto,gpu_memory_utilization=0.9,trust_remote_code=True,max_model_len=16384 --task gsm8k --num_fewshot 5  --batch_size auto
```

**Comment by kaixih** [comment]
> Edit: oh and ideally id still like to see accuracy numbers...

@LucasWilkinson this `DeepSeek-V2-Lite-Chat` only has attention head number == 16 and --tp=2 is not ok. Any advice on a model using head num == 128?

**Comment by kaixih** [comment]
> other than I do think we should turn it on by default for Blackwell, Any reason not to?

My main concern is that the CUTLASS MLA kernel has more limited support compared to the Triton version. For example, `num_heads` must be 128, and the page table must be padded to a multiple of `128 / page_size`. We're working on expanding the supported use cases, but until then, we'd prefer to keep this as an optional backend. What do you think?

**Comment by LucasWilkinson** [comment]
> > other than I do think we should turn it on by default for Blackwell, Any reason not to?
> 
> My main concern is that the CUTLASS MLA kernel has more limited support compared to the Triton version. For example, `num_heads` must be 128, and the page table must be padded to a multiple of `128 / page_size`. We're working on expanding the supported use cases, but until then, we'd prefer to keep this as an optional backend. What do you think?

Oh interesting, can we turn it on my default for models with head dim 128? as for block size I think we can force it similar to to FlashMLA (https://github.com/vllm-project/vllm/blob/5e6f93948449b8095e8eef5c3d99a8726e216a44/vllm/platforms/cuda.py#L145-L158), we plan to eventually handle this more gracefully (i.e. better support of the attention backend being able to specify the block size it wants if the user has not specified one)

Im ok landing this for now and opening a new PR to turn it on by default; but I think we still should, we should strive to make the best they can be.

**Comment by LucasWilkinson** [comment]
> > Edit: oh and ideally id still like to see accuracy numbers...
> 
> @LucasWilkinson this `DeepSeek-V2-Lite-Chat` only has attention head number == 16 and --tp=2 is not ok. Any advice on a model using head num == 128?

Ah the only one I know of is DeepSeek V3 or R1, if you have a system that can run that then those accuracy numbers would be good to see!

**Comment by kaixih** [comment]
Got the lm_eval output with the cutlass backend, which matches the triton backend:

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|_  | 0.96|_  |0.0197|
|     |       |strict-match    |     5|exact_match|_  | 0.96|_  |0.0197|
```

How to repro:
```bash
# server side: bash run.sh [cutlass]
if [[ "$1" == "cutlass" ]]; then
  export VLLM_ATTENTION_BACKEND=CUTLASS_MLA_VLLM_V1
fi

VLLM_LOGGING_LEVEL=DEBUG \
VLLM_WORKER_MULTIPROC_METHOD=spawn \
  vllm serve deepseek-ai/DeepSeek-V3 \
    --trust-remote-code \
    --max-model-len=4096 \
    --block-size=128 \
    --max-num-seqs=512 \
    --gpu_memory_utilization=0.97 \
    --data-parallel-size $NUM_GPUS --enable-expert-parallel \
    --disable-log-requests &

# client side
lm_eval --model local-completions --tasks gsm8k \
    --model_args model=deepseek-ai/DeepSeek-V3,base_url=http://127.0.0.1:8000/v1/completions,num_concurrent=5,max_retries=3,tokenized_requests=False \
    --limit 100
```

**Comment by kaixih** [comment]
The updated throughput after rebasing:

```
triton:
============ Serving Benchmark Result ============
Successful requests:                     3000
Benchmark duration (s):                  847.99
Total input tokens:                      2997000
Total generated tokens:                  3000000
Request throughput (req/s):              3.54
Output token throughput (tok/s):         3537.76
Total Token throughput (tok/s):          7071.98
---------------Time to First Token----------------
Mean TTFT (ms):                          155845.04
Median TTFT (ms):                        157803.49
P99 TTFT (ms):                           304408.28
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          686.05
Median TPOT (ms):                        685.18
P99 TPOT (ms):                           824.89
---------------Inter-token Latency----------------
Mean ITL (ms):                           686.05
Median ITL (ms):                         562.49
P99 ITL (ms):                            6502.67
==================================================

cutlass:
============ Serving Benchmark Result ============
Successful requests:                     3000
Benchmark duration (s):                  669.38
Total input tokens:                      2997000
Total generated tokens:                  3000000
Request throughput (req/s):              4.48
Output token throughput (tok/s):         4481.75
Total Token throughput (tok/s):          8959.01
---------------Time to First Token----------------
Mean TTFT (ms):                          154187.64
Median TTFT (ms):                        156351.06
P99 TTFT (ms):                           302269.57
-----Time per Output Token (excl. 1st token)------                                                                                                                                             Mean TPOT (ms):                          509.17
Median TPOT (ms):                        507.58                                                                                                                                                P99 TPOT (ms):                           649.29
---------------Inter-token Latency----------------
Mean ITL (ms):                           509.17
Median ITL (ms):                         370.06
P99 ITL (ms):                            6451.99
==================================================

```

**Comment by kaixih** [comment]
> we plan to eventually handle this more gracefully (i.e. better support of the attention backend being able to specify the block size it wants if the user has not specified one)
Im ok landing this for now and opening a new PR to turn it on by default; but I think we still should, we should strive to make the best they can be.

I agree — let's land this PR first, and then improve the attention backend picker in a more graceful way. I briefly looked into the code, and supporting this properly would require significant changes — for example, the function would need access to `model_config` (for num_heads), `cache_config` (for block_size), and `parallel_config` (for TP/DP). I think that level of change is better handled in a separate PR.

File: vllm/v1/attention/backends/mla/common.py [hunk 5]
```diff
@@ -349,7 +349,7 @@ def __init__(self,
         self.num_heads = model_config.get_num_attention_heads(
             runner.parallel_config)
         self.mla_dims = get_mla_dims(model_config)
-        self.aot_schedule = is_vllm_fa and (get_flash_attn_version() == 3)
+        self.aot_schedule = current_platform.is_cuda()
```
Comment (tlrmchlsmth) [inline vllm/v1/attention/backends/mla/common.py:352]: Is this right? I thought we still needed to guard against using `aot_schedule` if using FA2

Comment (kaixih) [reply vllm/v1/attention/backends/mla/common.py:352]: See comment [here](https://github.com/vllm-project/vllm/pull/17625#discussion_r2076437871). My understanding is that this aot_schedule is irrelevant here. @LucasWilkinson can correct me.

Comment (LucasWilkinson) [reply vllm/v1/attention/backends/mla/common.py:352]: Ya this is right, page-size is used by chunked-prefill in MLA, this naming was a bug

Comment (kaixih) [reply vllm/v1/attention/backends/mla/common.py:352]: Do I need to make any further changes, or is this good to go now?

**Review by tlrmchlsmth — APPROVED** [review]

**Comment by tlrmchlsmth** [comment]
This PR is pretty red... these tests aren't all failing on main are they?

**Comment by kaixih** [comment]
> This PR is pretty red... these tests aren't all failing on main are they?

I am checking the logs but nothing seems to be related to my change. Can you please advise? @tlrmchlsmth

**Comment by kaixih** [comment]
@tlrmchlsmth ok. I think i have done all I can do. Basically I have reverted all my changes that are for the existing files (only leave one that adding a new file). And you can see the tests are still red. Can you help?

**Comment by kaixih** [comment]
Force push the original PR.

**Comment by kaixih** [comment]
@tlrmchlsmth Can you please advise? ^^

**Comment by yuho8818** [comment]
did it support deepseek r1 ? I meet so error on deepseek r1

**Comment by kaixih** [comment]
@houseroad Thanks, I've just rebased. However, the `Lint and Deploy Charts / lint-and-deploy (pull_request)` check is failing, and it appears to be unrelated to the changes in this PR. Can you advise?
